src/VisualStudio/Core/Def/Implementation/Workspace/Esent/EsentPersistentStorage.cs

```diff
@@ -365,48 +320,76 @@ private static void WriteToStream(Stream inputStream, Stream esentStream, Cancel
             }
         }
 
-        private bool TryGetProjectAndDocumentId(Document document, out int projectId, out int documentId)
+        private EsentStorage.ProjectDocumentTableAccessor GetAccessor(EsentStorage.Key key)
         {
-            projectId = default(int);
-            documentId = default(int);
+            return key.DocumentIdOpt.HasValue ?
+                (EsentStorage.ProjectDocumentTableAccessor)_esentStorage.GetDocumentTableAccessor() :
+                (EsentStorage.ProjectDocumentTableAccessor)_esentStorage.GetProjectTableAccessor();
+        }
+
+        private bool TryGetProjectAndDocumentKey(Document document, out EsentStorage.Key key)
+        {
+            key = default(EsentStorage.Key);
+
+            int projectId;
+            int projectNameId;
+            int documentId;
+            if (!TryGetProjectId(document.Project, out projectId, out projectNameId) ||
+                !TryGetUniqueFileId(document.FilePath, out documentId))
+            {
+                return false;
+            }
+
+            key = new EsentStorage.Key(projectId, projectNameId, documentId);
+            return true;
+        }
 
-            if (!IsSupported(document.Project))
+        private bool TryGetProjectKey(Project project, out EsentStorage.Key key)
+        {
+            key = default(EsentStorage.Key);
+
+            int projectId;
+            int projectNameId;
+            if (!TryGetProjectId(project, out projectId, out projectNameId))
             {
                 return false;
             }
 
-            return TryGetUniqueFileId(document.Project.FilePath, out projectId) && TryGetUniqueFileId(document.FilePath, out documentId);
+            key = new EsentStorage.Key(projectId, projectNameId);
+            return true;
+        }
+
+        private bool TryGetProjectId(Project project, out int projectId, out int projectNameId)
+        {
```

Comment:
basically this is the change. I added projectNameId to (projectId, documentId) tuple.